### ISSUE TITLE
`tf.math.cumprod` gradient is incorrect when input contains zeros

### ISSUE THREAD

**Issue by wuyii8941** [tensorflow/tensorflow#119430]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

 TensorFlow: 2.13.1 (also reproducible on 2.16+)

### Custom code

Yes

### Summary

`tf.math.cumprod`'s gradient computation produces wrong values when any element of the input tensor is zero. The gradient for the zero element (and sometimes adjacent elements) is incorrectly returned as `0` instead of the mathematically correct value. This affects both eager mode and XLA.

### Root cause

The gradient implementation likely computes `d/dx[i]` using a formula involving `cumprod(x) / x[i]`. When `x[i] = 0`, this produces `0/0` because `cumprod[j]` for all `j >= i` is also `0`. The `0/0` evaluates to `NaN`, which is then silently replaced with `0`.

The correct approach is to compute the gradient using exclusive cumulative products from both directions (prefix and suffix), avoiding division entirely.

### Additional affected cases

| Input | TF gradient | Correct gradient | Wrong elements |
|-------|-------------|-----------------|----------------|
| `[0, 2, 3]` | `[0, 0, 0]` | `[9, 0, 0]` | `x[0]` |
| `[2, 3, 0]` | `[4, 2, 0]` | `[4, 2, 6]` | `x[2]` |
| `[2, 0, 0, 3]` | `[1, 0, 0, 0]` | `[1, 2, 0, 0]` | `x[1]` |

Note: when no zeros are present, the gradient is correct.

### Impact

Any training loop that uses `cumprod` on data that can contain zeros will silently receive wrong gradients, potentially causing the model to fail to learn or learn incorrectly. This includes survival analysis models (hazard functions), attention mechanisms with zero-masking, and probabilistic sequence models.

### Environment

- TensorFlow: 2.13.1 (also reproducible on 2.16+)
- OS: Ubuntu 20.04
- CPU-only (`CUDA_VISIBLE_DEVICES=""`)
- Also affects XLA (`jit_compile=True`)

### Standalone code to reproduce the issue

```shell
### Reproduction


import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
import tensorflow as tf
import numpy as np

x = tf.constant([1.5, 0.0, 2.0, 3.0], dtype=tf.float64)
with tf.GradientTape() as tape:
    tape.watch(x)
    y = tf.reduce_sum(tf.math.cumprod(x))
g = tape.gradient(y, x)
print("TF gradient:", g.numpy())       # [1. 0. 0. 0.]

# Numerical verification (central difference)
def f(xv):
    return float(np.sum(np.cumprod(xv)))
xv = np.array([1.5, 0.0, 2.0, 3.0])
num_g = np.zeros(4)
for i in range(4):
    xp, xm = xv.copy(), xv.copy()
    xp[i] += 1e-7; xm[i] -= 1e-7
    num_g[i] = (f(xp) - f(xm)) / 2e-7
print("Correct gradient:", num_g)       # [1. 13.5 0. 0.]


**Output:**

TF gradient: [1. 0. 0. 0.]
Correct gradient: [ 1.  13.5  0.   0. ]


The gradient for `x[1]` should be `13.5`, not `0`.

### Mathematical derivation

For `cumprod([x₀, x₁, x₂, x₃]) = [x₀, x₀x₁, x₀x₁x₂, x₀x₁x₂x₃]`:


d/dx₁ sum(cumprod) = d/dx₁ (x₀ + x₀x₁ + x₀x₁x₂ + x₀x₁x₂x₃)
                   = x₀ + x₀x₂ + x₀x₂x₃
                   = 1.5 + 1.5*2 + 1.5*2*3
                   = 1.5 + 3.0 + 9.0
                   = 13.5
```

### Relevant log output

```shell

```

**Comment by ChiragSW** [comment]
May I work on this issue?

**Comment by danyalahmed1995** [comment]
The issue is in the gradient path: the existing `Cumprod` gradient works for non-zero inputs, but the division-based path drops valid contributions when `x == 0`.

The patch keeps the existing path for non-zero elements and adds zero-specific handling for cumulative products that contain exactly one zero. Added coverage for zero positions, multiple zeros, `axis=1`, `exclusive`, `reverse`, and `exclusive + reverse`.
